stdlib/camlinternalFormat.ml

```diff
@@ -1412,11 +1412,29 @@ let format_of_fconv fconv prec =
     buffer_add_char buf symb;
     buffer_contents buf
 
+let transform_int_alt iconv s =
+  match iconv with
+  | Int_Cd | Int_Ci | Int_Cu ->
+    let buf = Buffer.create (String.length s * 4 / 3) in
+    let left = ref 0 in
+    for i = String.length s - 1 downto 0 do
+      match s.[i] with
+      | '0'..'9' -> if !left = 3 then left := 1 else incr left
+      | _ -> ()
+    done;
+    for i = 0 to String.length s - 1 do
+      match s.[i] with
+      | '0'..'9' as c -> if !left = 0 then (Buffer.add_char buf '_'; left := 3); decr left; Buffer.add_char buf c
+      | c -> Buffer.add_char buf c
+    done;
+    Buffer.contents buf
+  | _ -> s
+
```

Comment:
I think your idea of post-processing the C output is just fine. My suggestion remains in this style (as a `string -> string` function to apply on the returned value), but it allocates one less copy of the string. (I think some minds that work differently than mine may even consider it more readable.)